### PR TITLE
bpo-40521: [subinterpreters] Make dtoa bigint free list per-interpreter

### PR DESCRIPTION
**Pull request by dexterhahaha**
https://bugs.python.org/issue40521 
Make dtoa bigint free list per-interpreter

https://bugs.python.org/issue40521

### CODE REVIEW

File: Include/internal/pycore_dtoa.h [hunk 1]
```diff
@@ -17,7 +19,22 @@ PyAPI_FUNC(void) _Py_dg_freedtoa(char *s);
 PyAPI_FUNC(double) _Py_dg_stdnan(int sign);
 PyAPI_FUNC(double) _Py_dg_infinity(int sign);
 
+#define Kmax 7
+
+typedef uint32_t ULong;
+typedef int32_t Long;
+typedef uint64_t ULLong;
```
Comment (vstinner) [inline Include/internal/pycore_dtoa.h]: Many C files include directly or indirectly pycore_interp.h which includes pycore_dtoa.h. I would prefer to not pollule their namespaces with names like Kmax, ULong, Long, UULong which can clash with other names.

I see different options:

* Don't fix dtoa.c for now
* Add a prefix to all these names in the header file (ex: "_PyDtoa_") and create aliases in dtoa.c (ex: `#define Kmax _PyDtoa_Kmax`).
* Add an API like PyInterpreterState.dict to have per-interpreter variables: the freelist would be dynamically allocated on the heap memory, and stored as a raw pointer in PyInterpreterState. Or if you don't want to add a whole API, just expose the cache as a "void *" in the structure, and add init/fini functions to allocate/free this structure (free list).

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Include/internal/pycore_dtoa.h (hunk 1 above)
Comment (dexterhahaha) [reply Include/internal/pycore_dtoa.h]: thanks for your review. I chose the second option to fix this problem.

**Comment by JunyiXie** [comment]
I have made the requested changes; please review again

**Comment by JunyiXie** [comment]
I have removed the news. But I don't know how to make bedevere/news — No news entry in Misc/NEWS.d/next/ or "skip news" label found check success.

File: Include/internal/pycore_dtoa.h [hunk 2]
```diff
@@ -17,7 +19,22 @@ PyAPI_FUNC(void) _Py_dg_freedtoa(char *s);
...
+typedef uint64_t _PyDtoa_ULLong;
+
+struct
+Bigint {
+    struct Bigint *next;
+    int k, maxwds, sign, wds;
+    _PyDtoa_ULong x[1];
+};
+
+typedef struct Bigint Bigint;
```
Comment (vstinner) [inline Include/internal/pycore_dtoa.h]: This typedef is not needed in pycore_interp.h, please move it into dtoa.c.

File: Include/internal/pycore_dtoa.h [hunk 3]
```diff
@@ -17,7 +19,22 @@ PyAPI_FUNC(void) _Py_dg_freedtoa(char *s);
...
 
+#define _PyDtoa_Kmax 7
+
+typedef uint32_t _PyDtoa_ULong;
+typedef int32_t _PyDtoa_Long;
+typedef uint64_t _PyDtoa_ULLong;
+
+struct
+Bigint {
+    struct Bigint *next;
```
Comment (vstinner) [inline Include/internal/pycore_dtoa.h]:
```suggestion
struct _PyDtoa_Bigint {
    struct _PyDtoa_Bigint *next;
```

File: Include/internal/pycore_interp.h [hunk 4]
```diff
@@ -321,6 +322,7 @@ struct _is {
 
     struct ast_state ast;
     struct type_cache type_cache;
+    Bigint *bigint_freelist[_PyDtoa_Kmax+1];
```
Comment (vstinner) [inline Include/internal/pycore_interp.h]:
```suggestion
    struct _PyDtoa_Bigint *dtoa_freelist[_PyDtoa_Kmax + 1];
```

File: Python/dtoa.c [hunk 5]
```diff
@@ -361,9 +351,10 @@ Balloc(int k)
     int x;
     Bigint *rv;
     unsigned int len;
-
-    if (k <= Kmax && (rv = freelist[k]))
-        freelist[k] = rv->next;
+    PyInterpreterState *is = PyInterpreterState_Get();
```
Comment (vstinner) [inline Python/dtoa.c]: I would prefer to add a helper function get_freelist().

I suggest to use _PyInterpreterState_GET() which is faster than PyInterpreterState_Get().

File: Python/dtoa.c [hunk 6]
```diff
@@ -119,6 +119,12 @@
 
 #include "Python.h"
 #include "pycore_dtoa.h"
+#include "pycore_interp.h"
+
+#define Kmax _PyDtoa_Kmax
+#define ULong _PyDtoa_ULong
+#define Long _PyDtoa_Long
+#define ULLong _PyDtoa_ULLong
```
Comment (vstinner) [inline Python/dtoa.c]:
```suggestion
#define ULong _PyDtoa_ULong
#define Long _PyDtoa_Long
#define ULLong _PyDtoa_ULLong
#define Kmax _PyDtoa_Kmax

typedef struct _PyDtoa_Bigint Bigint;
```

File: Include/internal/pycore_dtoa.h [hunk 7]
```diff
@@ -1,4 +1,6 @@
 #ifndef PY_NO_SHORT_FLOAT_REPR
+#ifndef Py_CORE_DTOA_H
```
Comment (vstinner) [inline Include/internal/pycore_dtoa.h]: To follow other files, you should use "INTERNAL" instead: `#ifndef Py_INTERNAL_DTOA_H`.

File: Include/internal/pycore_interp.h (hunk 4 above)
Comment (vstinner) [inline Include/internal/pycore_interp.h]: You need to add:

```
#ifndef PY_NO_SHORT_FLOAT_REPR
...
#endif 
```

If the PY_NO_SHORT_FLOAT_REPR macro is defined, dtoa.c is not used.

**Comment by JunyiXie** [comment]
Thank vstinner, for your very detailed review. I did not consider these details before.

**Comment by JunyiXie** [comment]
I have made the requested changes; please review again

File: Python/dtoa.c [hunk 8]
```diff
@@ -351,7 +346,13 @@ typedef struct Bigint Bigint;
    Bfree to PyMem_Free.  Investigate whether this has any significant
    performance on impact. */
 
-static Bigint *freelist[Kmax+1];
+
+/* Get Bigint freelist from interpreter  */
+static Bigint **
+get_freelist() {
```
Comment (vstinner) [inline Python/dtoa.c]: See "function declaration isn’t a prototype [-Wstrict-prototypes]" compiler warning:
```suggestion
get_freelist(void) {
```

Comment (dexterhahaha) [reply Python/dtoa.c]: Thank you for your review, sorry I didn't pay attention to the warning.

**Comment by JunyiXie** [comment]
I have made the requested changes; please review again

**Comment by vstinner** [comment]
Can you please build Python with PY_NO_SHORT_FLOAT_REPR macro defined? For example, add -DPY_NO_SHORT_FLOAT_REPR=1 flag in CFLAGS.

**Comment by JunyiXie** [comment]
@vstinner  Thank you for your reminder, I added `-DPY_NO_SHORT_FLOAT_REPR=1` to run the unit test.
```
11 tests failed again:
     test_builtin test_configparser test_ctypes test_difflib test_float
     test_format test_fstring test_optparse test_statistics test_types
     test_unicode
```
```
======================================================================
FAIL: test_defaults_keyword (test.test_configparser.ConfigParserTestCase)
bpo-23835 fix for ConfigParser
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xiejunyi/bdcpython/Lib/test/test_configparser.py", line 979, in test_defaults_keyword
    self.assertEqual(cf[self.default_section]['1'], '2.4')
AssertionError: '2.3999999999999999' != '2.4'
- 2.3999999999999999
+ 2.4
======================================================================
FAIL: test_defaults_keyword (test.test_configparser.ConfigParserTestCaseLegacyInterpolation)
bpo-23835 fix for ConfigParser
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xiejunyi/bdcpython/Lib/test/test_configparser.py", line 979, in test_defaults_keyword
    self.assertEqual(cf[self.default_section]['1'], '2.4')
AssertionError: '2.3999999999999999' != '2.4'
- 2.3999999999999999
+ 2.4
```
After I revert my commit and run the unit test as well, there will still be these errors.
I guess defining `PY_NO_SHORT_FLOAT_REPR=1` will cause these errors.  not my commit cause those test faileds.

**Comment by vstinner** [comment]
I merged your PR thanks.

> @vstinner Thank you for your reminder, I added -DPY_NO_SHORT_FLOAT_REPR=1 to run the unit test.

I just wanted to check if it's possible to build Python in the PY_NO_SHORT_FLOAT_REPR macro defined: it's the case, thanks. I'm not interetesed by test_configparser failures. Platforms where PY_NO_SHORT_FLOAT_REPR macro is defined are rare. If an user of one of such platform is impacted, they can report an issue.

**Pull request merged**